diff -ruN retrace-server-1.12.ori/src/config/retrace-server.conf retrace-server-1.12.new/src/config/retrace-server.conf
--- retrace-server-1.12.ori/src/config/retrace-server.conf	2015-04-07 22:38:25.753999980 +0800
+++ retrace-server-1.12.new/src/config/retrace-server.conf	2015-04-07 22:39:52.791999809 +0800
@@ -8,19 +8,19 @@
 AuthGroup = retrace
 
 # Force to use HTTPS - only disable on trusted network
-RequireHTTPS = 1
+RequireHTTPS = 0
 
 # Allow to delete task data via HTTP API (https://server/<task_id>/delete)
 AllowAPIDelete = 0
 
 # Allow interactive tasks (security risk, do not use on public systems)
-AllowInteractive = 0
+AllowInteractive = 1
 
 # Allow X-CoreFileDirectory header
 AllowExternalDir = 0
 
 # Allow to create tasks owned by task manager (security risk)
-AllowTaskManager = 0
+AllowTaskManager = 1
 
 # If white list is disabled, anyone can delete tasks
 TaskManagerAuthDelete = 0
@@ -75,10 +75,10 @@
 
 # Local repos directory
 # if changed, you also need to update httpd config
-RepoDir = /var/cache/retrace-server
+RepoDir = /cores/retrace/repos
 
 # Directory where the crashes and results are saved
-SaveDir = /var/spool/retrace-server
+SaveDir = /cores/retrace/spool
 
 # Directory where old tasks are moved
 DropDir = /srv/retrace/archive
diff -ruN retrace-server-1.12.ori/src/config/retrace-server-httpd.conf retrace-server-1.12.new/src/config/retrace-server-httpd.conf
--- retrace-server-1.12.ori/src/config/retrace-server-httpd.conf	2015-04-07 22:38:25.753999980 +0800
+++ retrace-server-1.12.new/src/config/retrace-server-httpd.conf	2015-04-07 22:40:20.105001727 +0800
@@ -15,7 +15,8 @@
 WSGIScriptAliasMatch ^/[0-9]+/start$ /usr/share/retrace-server/start.wsgi
 WSGIScriptAliasMatch ^/$ /usr/share/retrace-server/index.wsgi
 
-<Directory "/var/cache/retrace-server">
+#<Directory "/var/cache/retrace-server">
+<Directory "/cores/retrace/repos">
     Options Indexes FollowSymLinks
     AllowOverride None
     <IfModule mod_authz_core.c>
@@ -24,7 +25,7 @@
     </IfModule>
     <IfModule !mod_authz_core.c>
         # Apache 2.2
-        Order allow,deny
+       Order allow,deny
         Allow from all
     </IfModule>
 </Directory>
@@ -42,4 +43,5 @@
     </IfModule>
 </LocationMatch>
 
-Alias /repos /var/cache/retrace-server
+#Alias /repos /var/cache/retrace-server
+Alias /repos /cores/retrace/repos
